plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/DefaultThriftMetastoreClientFactory.java

```diff
@@ -97,14 +124,56 @@ public DefaultThriftMetastoreClientFactory(
                 Optional.ofNullable(config.getSocksProxy()),
                 config.getMetastoreTimeout(),
                 metastoreAuthentication,
-                nodeManager.getCurrentNode().getHost());
+                nodeManager.getCurrentNode().getHost(),
+                buildThriftHttpContext(httpMetastoreConfig),
+                openTelemetry);
     }
 
     @Override
-    public ThriftMetastoreClient create(HostAndPort address, Optional<String> delegationToken)
+    public ThriftMetastoreClient create(URI uri, Optional<String> delegationToken)
             throws TTransportException
     {
-        return create(() -> createTransport(address, delegationToken), hostname);
+        return create(() -> getTransportSupplier(uri, delegationToken), hostname);
+    }
+
+    private TTransport getTransportSupplier(URI uri, Optional<String> delegationToken)
+            throws TTransportException
+    {
+        switch (uri.getScheme().toLowerCase(ENGLISH)) {
+            case "thrift" -> {
+                return createTransport(HostAndPort.fromParts(uri.getHost(), uri.getPort()), delegationToken);
+            }
+            case "http", "https" -> {
+                return createHttpTransport(uri, thriftHttpContext.orElseThrow(() -> new IllegalArgumentException("Thrift http context is not set")));
+            }
+            default -> throw new IllegalArgumentException("Invalid metastore uri scheme " + uri.getScheme());
+        }
+    }
+
+    private TTransport createHttpTransport(URI uri, ThriftHttpContext httpThriftContext)
+            throws TTransportException
+    {
+        HttpClientBuilder clientBuilder = createHttpClientBuilder(httpThriftContext);
+        clientBuilder.setDefaultRequestConfig(RequestConfig.custom().setConnectTimeout(timeoutMillis).build());
+        return new THttpClient(uri.toString(), clientBuilder.build());
+    }
+
+    private HttpClientBuilder createHttpClientBuilder(ThriftHttpContext httpThriftContext)
+    {
+        HttpClientBuilder httpClientBuilder = ApacheHttpClientTelemetry.builder(openTelemetry).build().newHttpClientBuilder();
+        if (sslContext.isPresent()) {
+            SSLConnectionSocketFactory socketFactory = new SSLConnectionSocketFactory(sslContext.get(), new DefaultHostnameVerifier(null));
+            Registry<ConnectionSocketFactory> registry = RegistryBuilder
+                    .<ConnectionSocketFactory>create()
+                    .register("https", socketFactory)
+                    .build();
+            httpClientBuilder.setConnectionManager(new BasicHttpClientConnectionManager(registry));
+        }
+        httpClientBuilder.addInterceptorFirst((HttpRequestInterceptor) (httpRequest, httpContext) -> {
+            httpRequest.addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + httpThriftContext.token());
```

Comment:
I'm a little wondering if we want to allow bearer auth even when the URI is http (not https) from security perspective.

Reply:
Tweaking validation just for tests doesn't make sense to me.